packages/aila/src/lib/agents/promptAgentHandler.ts

```diff
@@ -19,54 +25,150 @@ export async function promptAgentHandler<Schema extends SchemaWithValue>({
   userId,
   ragData,
 }: {
-  agent: PromptAgentDefinition<Schema>;
+  agent: PromptAgentDefinition<Schema, SchemaStrict>;
   document: LooseLessonPlan;
   targetKey: LessonPlanKey;
   additionalInstructions: string;
   chatId: string;
   userId: string;
   ragData: CompletedLessonPlan[];
-}): Promise<{ content: z.infer<Schema>["value"] }> {
+}): Promise<
+  | { success: true; content: z.infer<SchemaStrict> }
+  | { success: false; error: string }
+> {
   const openAIClient = createOpenAIClient({
     app: "lesson-assistant",
     chatMeta: {
       chatId,
       userId,
     },
   });
-  const responseFormat = zodTextFormat(
-    agent.schema,
-    `${agent.name}_response_schema`,
-  );
+  const schema = z.object({
+    output: z.union([
+      z.object({ success: z.literal(true), value: agent.schemaForLLM }),
+      z.object({ success: z.literal(false), error: z.string() }),
+    ]),
+  });
+  const responseFormat = zodTextFormat(schema, `${agent.name}_response_schema`);
 
-  const result = await openAIClient.responses.parse({
-    instructions: agent.prompt,
-    input: `### Examples from similar lessons
+  const instructions = agent.prompt;
+  const input = `### Examples from similar lessons
 
-${ragData.map((lesson) => `#### ${targetKey} ${lesson.title}\n\n${agent.extractRagData(lesson)}`).join("\n\n")}
+${ragData.map((lesson, i) => `#### Example ${i + 1}: ${targetKey} from lesson '${lesson.title}'\n\n${agent.extractRagData(lesson)}`).join("\n\n")}
```

Comment:
Will add some more colour to this in [follow up work](https://www.notion.so/oaknationalacademy/Add-prompt-detail-for-based-on-content-ensure-it-s-properly-injected-and-improve-on-relevant-le-1fc26cc4e1b18041a03bd8714a9cfa40?source=copy_link)